### PR TITLE
feat: enable metasrv leader cached kv

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Enable metasrv leader cached kv

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/meta-srv/src/metasrv.rs [hunk 1]
```diff
@@ -176,17 +178,29 @@ pub struct MetaStateHandler {
...
 
 impl MetaStateHandler {
     pub async fn on_become_leader(&self) {
+        self.state.write().unwrap().next_state(become_leader(false));
+
+        if let Err(e) = self.leader_cached_kv_store.load().await {
+            error!(e; "Failed to load kv into leader cache kv store");
+        } else {
+            self.state.write().unwrap().next_state(become_leader(true));
+        }
```
Comment (TheWaWaR) [inline src/meta-srv/src/metasrv.rs:193]: What if `self.state.write().unwrap().next_state(become_follower());` is called in other threads when `on_become_leader` is processing?  Lock `self.state` this way, I'm afraid there will be some data races in the future. I suggest use `self.state` to lock the entire function.

```rust
pub async fn on_become_leader(&self) {
    let state = self.state.write().unwrap();
    if let Err(e) = self.leader_cached_kv_store.load().await {
        error!(e; "Failed to load kv into leader cache kv store");
        state.next_state(become_leader(false));
    } else {
        state.next_state(become_leader(true));
    }
    // other logic...
}
```

This logic is also applied to the usage of `self.state` in `on_become_follower()`.

An addition question: `StateRef` is not using anywhere else in the code base, so what's the point of introducing this `struct State`?

Comment (WenyXu) [reply src/meta-srv/src/metasrv.rs:193]: In this scenario, we only use a single thread.

File: src/meta-srv/src/service/store/cached_kv.rs [hunk 2]
```diff
@@ -79,6 +92,38 @@ impl LeaderCachedKvStore {
         Self::new(Arc::new(AlwaysLeader), store)
     }
 
+    /// The caller MUST ensure during the loading, there are no mutation requests reaching the `LeaderCachedKvStore`.
```
Comment (MichaelScofield) [inline src/meta-srv/src/service/store/cached_kv.rs:95]: Can the caller actually ensure this? The loading of leader cache is executed in a bg task, while the other parts in Metasrv is running at the same time, changing the etcd simultaneously.

Comment (WenyXu) [reply src/meta-srv/src/service/store/cached_kv.rs:95]: Currently, All mutations will be executed in Metasrv's procedures.

**Comment by WenyXu** [comment]
@MichaelScofield @fengjiachun PTAL

**Review by MichaelScofield — APPROVED** [review]

File: src/meta-srv/src/handler/on_leader_start_handler.rs [hunk 3]
```diff
@@ -1,44 +0,0 @@
...
-
-    async fn handle(
-        &self,
-        _req: &HeartbeatRequest,
-        ctx: &mut Context,
-        _acc: &mut HeartbeatAccumulator,
-    ) -> Result<()> {
-        if let Some(election) = &ctx.election {
-            if election.in_infancy() {
-                ctx.is_infancy = true;
```
Comment (fengjiachun) [inline src/meta-srv/src/handler/on_leader_start_handler.rs:37]: We still need the field: `is_infancy`

**Review by fengjiachun — APPROVED** [review]
LGTM

**Pull request merged**